docs for external data (#1677)
open-policy-agent/gatekeeper: website/docs/externaldata.md

---
id: externaldata
title: External Data
---

> This feature is still in alpha stage, so the final form can still change (feedback is welcome!). Mutation is not yet supported with external data.

## Motivation

Gatekeeper provides various means to mutate and validate Kubernetes resources. However, in many of these scenarios this data is either built-in, static or user-defined. With external data feature, we are enabling Gatekeeper to interface with various external data sources, such as image registries, using a provider-based model.

A similar way to connect with an external data source can be done today using OPA's built-in `http.send` functionality. However, there are limitations to this approach.
- Gatekeeper does not support rego policies for mutation, which can not use the OPA http.send built-in function

> - There are unknown factors when the proxy endpoint is not responsive, doesn't exist/down or fails. These are addressed by external data interfaces such as timeout.

Reviewer: Also, potential security risks... if you don't trust the author of a constraint template, then `http.Send()` is a bad idea, as it gives them network access.

If you believe the CT author is safe, they still need to carefully consider how their Rego is written to avoid injection attacks. External Data helps with this by:

* Restricting which hosts a user can access
* Providing an interface for making requests, which allows us to better handle things like escaping strings.

Also, potential for performance benefits, as we can more directly control caching. In the case of mutation we can batch requests via lazy evaluation. For validation we make batching easier via function design.

Finally, using external data primitives makes it easier to keep track of data sources by providing a common format for registering and accessing them.

Those are the other benefits that come to mind. I may be missing some.

I'd avoid listing failure handling/timeouts/etc. as a way `http.Send()` doesn't fit, since it could probably be extended to provide these. The biggest differentiator is that `http.Send()` will send any request to any host, whereas external data wont (along with other restrictions on how the request is formatted/made). This makes it less flexible, but also provides us space to offer value as a platform, like mitigating some of the above security concerns.

Author: Thank you! updated

Using the external data solution provides benefits, such as:

> - Tackle common scenarios together, such as image SHA mutation that many of these scenarios will utilize.

Reviewer: I'm not sure what "tackle common scenarios together" means? I realize that SHA mutation is a common scenario, but what does it mean to "tackle it together"?

Author: Changed this a bit. Lmk what you think

> - Provider based model provides examples and guidelines for each provider to integrate with Gatekeeper

Reviewer: Provider model creates a common interface for extending Gatekeeper with external data.

This means that developers and consumers of data sources can rely on that common protocol to ease authoring of both constraint templates and data sources. It also makes change management a bit easier... if a data source is compliant with external data, users of that source should be able to tell whether upgrading it will break existing CTs (and our goal is to have that answer always be "no").

It also allows us to provide "free" value as a platform, like the performance benefits for mutation.

Author: Thank you! updated.

---
id: externaldata
title: External Data
---

> > This feature is still in alpha stage, so the final form can still change (feedback is welcome!). Mutation is not yet supported with external data.

Reviewer 2: ```suggestion
> ❗ This feature is still in alpha stage, so the final form can still change (feedback is welcome!). 
> 🚧  Mutation is not yet supported with external data.
```

Author: updated

## Motivation

Gatekeeper provides various means to mutate and validate Kubernetes resources. However, in many of these scenarios this data is either built-in, static or user defined. With external data feature, we are enabling Gatekeeper to interface with various external data sources, such as image registries, using a provider-based model.

A similar way to connect with an external data source can be done today using OPA's built-in `http.send` functionality. However, there are limitations to this approach.
- Gatekeeper does not support Rego policies for mutation, which cannot use the OPA `http.send` built-in function.
- Security concerns due to:
  - if template authors are not trusted, it will potentially give template authors access to the in-cluster network.
  - if template authors are trusted, authors will need to be careful on how rego is written to avoid injection attacks.

> Using the external data solution provides benefits, such as:

Reviewer 2: ```suggestion
Key benefits provided by the external data solution:
```

Author: updated

- Addresses security concerns by:
  - Restricting which hosts a user can access.

>   - Providing an interface for making requests, which allows us to better handle things like escaping strings.

Reviewer: nit: "us" -> "Gatekeeper"

> - Addresses common scenarios with a single provider, such as image SHA mutation that many of these scenarios may utilize.

Reviewer: Are we using the word "scenarios" in two different ways on this bullet?

Author: updated

> - Addresses common patterns with a single provider, such as image SHA mutation, that potential scenarios (such as validating image signatures or vulnerabilities) may utilize.

Reviewer 2: ```suggestion
- Addresses common patterns with a single provider, e.g. image tag to digest mutation, which can be leveraged by multiple scenarios (e.g. validating image signatures or vulnerabilities).
```

Author: updated

> - Provider model creates a common interface for extending Gatekeeper with external data.

Reviewer 2: another one:

Provider model allows for separation of concerns between the implementation that allows access to external data and the actual policy being evaluated.

Author: updated

  - Developers and consumers of data sources can rely on that common protocol to ease authoring of both constraint templates and data sources.

>   - Makes change management easier as users of an external data provider should be able to tell whether upgrading it will break existing constraint templates. (our goal is to have that answer always be "no")

Reviewer: "our goal is to have that answer"... maybe we want to re-word this a bit. That is certainly our intent long term, but maybe not while we're in alpha. Alpha is about figuring out a stable interface that is likely to endure.

Author: clarified with "once external data API is stable"

> - Performance benefits as we can now directly control caching.

Reviewer: Also, we get to control which keys are significant for caching, increasing the likelihood of cache hits.

Author: added

  - For mutation, we can batch requests via lazy evaluation.

>   - For validation, we make batching easier via function design.

Reviewer 2: > For validation, we make batching easier via function design

what does "function design" mean here?

Author: I think this means `external_data` rego function receives a list. @maxsmythe is this correct? should we clarify?

Reviewer: yep, the function signature makes it easier to batch calls in Rego by accepting a list.

Author: added clarification that it's for rego func

## Enabling external data support

### YAML
You can enable external data support by adding `--enable-external-data` in gatekeeper audit and controller-manager deployment arguments.

### Helm
You can also enable external data by installing or upgrading Helm chart by setting `enableExternalData=true`:

```sh
helm install gatekeeper/gatekeeper --name-template=gatekeeper --namespace gatekeeper-system --create-namespace \
	--set enableExternalData=true \

> 	--set controllerManager.dnsPolicy=ClusterFirst,audit.dnsPolicy=ClusterFirst

Reviewer 2: do we still need this if default is set to `ClusterFirst` now?

Author: we haven't released an update with this yet, v3.7.0 release in the Helm repo doesn't have this.

Reviewer 2: nevermind I see it's in manifest_staging only.

### Dev/Test

For dev/test deployments, use `make deploy ENABLE_EXTERNAL_DATA=true`

## Providers

Providers are designed to be in-cluster components that can communicate with external data sources (such as image registries, Active Directory/LDAP directories, etc) and return data in a format that can be processed by Gatekeeper.

Example provider can be found at: https://github.com/open-policy-agent/gatekeeper/tree/master/test/externaldata/dummy-provider

## Provider implementation

Provider is an HTTP server that listens on a port and responds to [ProviderRequest](#ProviderRequest) with [ProviderResponse](#ProviderResponse).

As part of [ProviderResponse](#ProviderResponse), the provider can return a list of items. Each item is a JSON object with the following fields:
- `Key`: the key that was sent to the provider
- `Value`: the value that was returned from the provider for that key
- `Error`: an error message if the provider returned an error for that key

If there is a system error, the provider should return the system error message in the `SystemError` field.

> > Recommendation is to keep a timeout such as maximum of 1-2 seconds for the provider to respond.

Reviewer 2: maybe consider moving this after the provider crd as before the crd, there's no mention of `timeout` field at all.

Author: I had it in the comment. added clarification

Example provider implementation: https://github.com/open-policy-agent/gatekeeper/blob/master/test/externaldata/dummy-provider/provider.go

## Gatekeeper Provider custom resource

Provider resource defines the provider and the configuration for it.

```yaml
apiVersion: externaldata.gatekeeper.sh/v1alpha1
kind: Provider
metadata:
  name: my-provider
spec:
  url: http://<service-name>.<namespace>:<port>/<endpoint> # URL to the external data source (e.g., http://my-provider.my-namespace:8090/validate)
  timeout: <timeout> # timeout value in seconds (e.g., 1)
```

> ## Gatekeeper validation

Reviewer 2: ```suggestion
## External data for Gatekeeper validating webhook
```

External data adds a [custom OPA built-in function](https://www.openpolicyagent.org/docs/latest/extensions/#custom-built-in-functions-in-go) called `external_data` to Rego. This function is used to query external data providers.

`external_data` is a function that takes a request and returns a response. The request is a JSON object with the following fields:
- `Provider`: the name of the provider to query
- `Keys`: the list of keys to send to the provider

```rego
  # send external data request
  response := external_data({"provider": "my-provider", "keys": my-list})
```

Response example: [[`"my-key"`, `"my-value"`, `""`], [`"another-key"`, `42`, `""`], [`"bad-key"`, `""`, `"error message"`]]

> > To avoid multiple calls to the provider, recommendation is to batch the keys list and send the request only once.

Reviewer 2: ```suggestion
> To avoid multiple calls to the same provider, recommendation is to batch the keys list to send a single the request.
```

Reviewer 2: might be good to provide an example of the batch request keys list.

Author: added
